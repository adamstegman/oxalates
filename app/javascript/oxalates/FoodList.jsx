import React from 'react';
import PropTypes from 'prop-types';

import { foodPropType } from './foodPropType';
import { FoodListEmptyItem } from './FoodListEmptyItem';
import { FoodListError } from './FoodListError';
import { FoodListItem } from './FoodListItem';
import { FoodListLoading } from './FoodListLoading';
import { FoodListNewItem } from './FoodListNewItem';
import { listPropType } from './listPropType';

import './FoodList.scss';

export class FoodList extends React.Component {
  // Find the "highest" (most dangerous) list
  getListForFood(food, lists) {
    return lists.find(list => (
      (
        (list.bottomThreshold && list.bottomThreshold <= food.oxalateMg) &&
        (list.topThreshold && list.topThreshold >= food.oxalateMg)
      ) ||
      (list.bottomThreshold === null && (list.topThreshold && list.topThreshold >= food.oxalateMg)) ||
      ((list.bottomThreshold && list.bottomThreshold <= food.oxalateMg) && list.topThreshold === null)
    ));
  }

  sortFoodsAlphabetically(foods) {
    return foods.sort((a, b) => a.name.localeCompare(b.name));
  }

  sortListsByThresholds(lists) {
    return lists.sort((a, b) => {
      const aBottom = a.bottomThreshold || 0;
      const aTop = a.topThreshold || Number.MAX_VALUE;
      const aTotal = aBottom + aTop;
      const bBottom = b.bottomThreshold || 0;
      const bTop = b.topThreshold || Number.MAX_VALUE;
      const bTotal = bBottom + bTop;
      if (aTotal < bTotal) {
        return 1;
      } else if (aTotal > bTotal) {
        return -1;
      } else {
        return 0;
      }
    });
  }

  render() {
    const {
      foods,
      lists,
      requestedListId,
      newFood,
      newFoodListId,
      error,
      password,
      createFood,
      setNewFood,
    } = this.props;
    const sortedFoods = this.sortFoodsAlphabetically(foods);
    const sortedLists = this.sortListsByThresholds(lists);

    let contents = <FoodListEmptyItem />;
    if (requestedListId) {
      contents = <FoodListLoading />;
    } else if (newFood) {
      contents = <FoodListNewItem newFood={newFood}
                                  newFoodListId={newFoodListId}
                                  createFood={createFood}
                                  setNewFood={setNewFood}
                                  error={error}
                                  password={password} />;
    } else if (error) {
      contents = <FoodListError error={error} />;
    } else if (foods.length > 0) {
      contents = sortedFoods.map(food => {
        const list = this.getListForFood(food, sortedLists);
        return <FoodListItem key={food.id} food={food} list={list} />;
      });
    }

    return (
      <div className="content">
        <ul className="food-list">
          {contents}
        </ul>
      </div>
    );
  }
};

FoodList.propTypes = {
  foods: PropTypes.arrayOf(foodPropType).isRequired,
  lists: PropTypes.arrayOf(listPropType).isRequired,
  requestedListId: PropTypes.node,
  error: PropTypes.string,
  newFood: PropTypes.shape({}),
  newFoodListId: PropTypes.node,
  password: PropTypes.string,
  createFood: PropTypes.func,
  setNewFood: PropTypes.func,
};
