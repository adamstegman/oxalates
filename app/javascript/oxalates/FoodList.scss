$food_list_item_border: 1px solid #ddd;

.food-list {
  list-style-type: none;
  margin: 0;
  padding: 0;

  & > li {
    display: -webkit-flex;
    display: flex;
    -webkit-align-items: center;
    align-items: center;

    border-bottom: $food_list_item_border;
  }
}
