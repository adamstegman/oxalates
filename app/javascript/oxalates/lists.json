[
  {
    "id": "all",
    "name": "All",
    "bottom_threshold": null,
    "top_threshold": null
  },
  {
    "id": 1,
    "name": "Very High",
    "bottom_threshold": 50,
    "top_threshold": null
  },
  {
    "id": 2,
    "name": "High",
    "bottom_threshold": 15,
    "top_threshold": 50
  },
  {
    "id": 3,
    "name": "Moderate",
    "bottom_threshold": 5,
    "top_threshold": 15
  },
  {
    "id": 4,
    "name": "Low",
    "bottom_threshold": null,
    "top_threshold": 5
  }
]
