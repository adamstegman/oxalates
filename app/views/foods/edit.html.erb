<%= form_for @food do |f| %>
  <%= f.label :name %>
  <%= f.text_field :name %>
  <%= f.label :list %>
  <%= f.select :list, options_from_collection_for_select(all_lists, :id, :name, @food.list_id) %>

  <%= f.submit "Save" %>
<% end %>
