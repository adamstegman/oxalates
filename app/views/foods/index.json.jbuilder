json.foods(@foods) do |food|
  json.(food, :id, :name, :oxalate_mg, :serving)
end
