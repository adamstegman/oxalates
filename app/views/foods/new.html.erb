<%= form_for @food, html: {class: 'edit-food'} do |f| %>
  <p><%= f.label :name, "New food name" %></p>
  <p><%= f.text_field :name %></p>
  <p><%= f.label :oxalate_mg, "Oxalates per serving (mg)" %></p>
  <p><%= f.number_field :oxalate_mg, step: "any" %></p>
  <p><%= f.label :serving %></p>
  <p><%= f.text_field :serving %></p>

  <p><%= f.submit "Add" %></p>
<% end %>
