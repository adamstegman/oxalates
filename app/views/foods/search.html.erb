<h1>Results</h1>

<ul>
	<% @results.each do |r| %>
		<li><%= r.name + ": " + r.list.name %></li> 
	<% end %>
</ul>