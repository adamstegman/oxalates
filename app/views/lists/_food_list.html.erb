<ul class="food-list">
<% foods.each do |food| %>
  <li class="food-list-item <%= list_class_name(food.list) %>"><%= food.name %></li>
<% end %>
</ul>
