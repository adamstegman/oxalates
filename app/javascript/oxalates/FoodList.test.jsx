import React from 'react';
import renderer from 'react-test-renderer';

import foods from './__mocks__/foods.json';
import { FoodList } from './FoodList';
import lists from './__mocks__/lists.json';

test('FoodList renders foods', () => {
  const component = renderer.create(
    <FoodList foods={foods} lists={lists} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodList renders empty food list', () => {
  const component = renderer.create(
    <FoodList foods={[]} lists={lists} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodList renders loading indicator', () => {
  const component = renderer.create(
    <FoodList foods={[]} lists={lists} requestedListId={lists[0].id} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodList renders new food form', () => {
  const component = renderer.create(
    <FoodList foods={foods}
              lists={lists}
              newFood={{}}
              newFoodListId={lists[0].id}
              password={'password'}
              createFood={() => {}}
              setNewFood={() => {}} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('FoodList renders error indicator', () => {
  const err = 'some error';
  const component = renderer.create(
    <FoodList foods={[]} lists={lists} error={err} />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
