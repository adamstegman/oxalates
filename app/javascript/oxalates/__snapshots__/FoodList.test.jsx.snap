// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodList renders edit food form 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <form
      className="edit-food"
      onSubmit={[Function]}
    >
      <label
        htmlFor="name"
      >
        Name
      </label>
      <input
        autoFocus={true}
        name="name"
        onChange={[Function]}
        type="text"
        value="Asparagus, raw"
      />
      <label
        htmlFor="oxalateMg"
      >
        Oxalates per serving (mg)
      </label>
      <input
        name="oxalateMg"
        onChange={[Function]}
        step="any"
        type="number"
        value={4.9}
      />
      <label
        htmlFor="serving"
      >
        Serving
      </label>
      <input
        name="serving"
        onChange={[Function]}
        type="text"
        value="1/2 cup"
      />
      <input
        type="submit"
        value="Update"
      />
    </form>
  </ul>
</div>
`;

exports[`FoodList renders editing foods 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <li
      className="editing-food-list-item"
    >
      <button
        className="destroy-food-list-item"
        onClick={[Function]}
        type="button"
      >
        <test-file-stub />
      </button>
      <button
        className="edit-food-list-item low"
        onClick={[Function]}
        type="button"
      >
        Asparagus, raw
         - 
        4.90
        mg - 
        1/2 cup
      </button>
    </li>
    <li
      className="editing-food-list-item"
    >
      <button
        className="destroy-food-list-item"
        onClick={[Function]}
        type="button"
      >
        <test-file-stub />
      </button>
      <button
        className="edit-food-list-item very-high"
        onClick={[Function]}
        type="button"
      >
        beetroot, steamed
         - 
        50.00
        mg - 
        just under 1/2 cup
      </button>
    </li>
    <li
      className="editing-food-list-item"
    >
      <button
        className="destroy-food-list-item"
        onClick={[Function]}
        type="button"
      >
        <test-file-stub />
      </button>
      <button
        className="edit-food-list-item high"
        onClick={[Function]}
        type="button"
      >
        carrots, steamed
         - 
        15.00
        mg - 
        1/2 cup
      </button>
    </li>
    <li
      className="editing-food-list-item"
    >
      <button
        className="destroy-food-list-item"
        onClick={[Function]}
        type="button"
      >
        <test-file-stub />
      </button>
      <button
        className="edit-food-list-item moderate"
        onClick={[Function]}
        type="button"
      >
        tomato, raw, sliced
         - 
        5.00
        mg - 
        1/2 cup
      </button>
    </li>
  </ul>
</div>
`;

exports[`FoodList renders empty food list 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <li
      className="food-list-item"
    >
      No foods found
    </li>
  </ul>
</div>
`;

exports[`FoodList renders error indicator 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <li
      className="food-list-item error"
    >
      some error
    </li>
  </ul>
</div>
`;

exports[`FoodList renders foods 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <li
      className="food-list-item low"
    >
      Asparagus, raw
       - 
      4.90
      mg - 
      1/2 cup
    </li>
    <li
      className="food-list-item very-high"
    >
      beetroot, steamed
       - 
      50.00
      mg - 
      just under 1/2 cup
    </li>
    <li
      className="food-list-item high"
    >
      carrots, steamed
       - 
      15.00
      mg - 
      1/2 cup
    </li>
    <li
      className="food-list-item moderate"
    >
      tomato, raw, sliced
       - 
      5.00
      mg - 
      1/2 cup
    </li>
  </ul>
</div>
`;

exports[`FoodList renders loading indicator 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <li
      className="food-list-item"
    >
      Loading foods...
    </li>
  </ul>
</div>
`;

exports[`FoodList renders new food form 1`] = `
<div
  className="content"
>
  <ul
    className="food-list"
  >
    <form
      className="add-food"
      onSubmit={[Function]}
    >
      <label
        htmlFor="name"
      >
        New food name
      </label>
      <input
        autoFocus={true}
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="oxalateMg"
      >
        Oxalates per serving (mg)
      </label>
      <input
        name="oxalateMg"
        onChange={[Function]}
        step="any"
        type="number"
        value={0}
      />
      <label
        htmlFor="serving"
      >
        Serving
      </label>
      <input
        name="serving"
        onChange={[Function]}
        type="text"
        value=""
      />
      <input
        type="submit"
        value="Add"
      />
    </form>
  </ul>
</div>
`;
